Druid Catalog basics

paul-rogers opened the pull request:
The [Druid catalog](https://github.com/apache/druid/issues/12546) provides a collection of metadata "hints" about tables (datasources, input sources, views, etc.) within Druid. This PR provides the foundation: the DB and REST layer, but not yet the integration with the Calcite SQL layer. This is a much-refined version of the [earlier catalog PR](https://github.com/apache/druid/pull/12647).

The DB layer extends what is done for other Druid metadata tables.  The semantic ("business logic") layer provides the usual CRUD operations on tables. The entire design is pretty standard and follows Druid patterns. The key difference is the rather extreme lengths taken by the implementation to ensure each bit is easily testable without mocks. That means many interfaces which can be implemented in multiple ways.

Parts available in this PR include:

* The metadata DB storage layer (in an extension)
* The basic "catalog object model" that describes the properties and columns which describe catalog tables.
* A basic set of tables: datasources and three kinds of external tables (inline, local, and HTTP).
* A REST API layer to perform CRUD operations on tables.
* Unit tests
* An integration test of the catalog REST API.

The catalog mechanism is split into two parts.

* The "core" part which describes catalog objects, and which is can model data from a variety of catalog systems.
* The `druid-catalog` extension which stores data in the Druid metadata database.

This split exists for two reasons:

* Many metadata systems exist: HMS, Amazon Glue, various commercial solutions, etc. We anticipate that some shops may wish to obtain metadata from these other systems, in the same way that some shops get their security information from external systems.
* Druid's database schema evolution system is rather basic. (And, by "basic", we mean "nonexistent.") There is some chance that the remaining development will change the schema, which upgrades cannot support. Users who enable `druid-catalog` extension now acknowledge that they are using it only for testing, not production, and at their own risk.

Functionality not in this PR, but which will appear in the next one, includes:

* The synchronization mechanism between the Coordinator and Broker.
* SQL table functions to make use of catalog entries.
* Integration of catalog properties to simplify MSQ ingest statements (`INSERT` and `REPLACE`).
* Integration of catalog schema information with `SELECT` queries.
* The remaining set of external table types.
* Views.

This is a great opportunity for reviewers to provide guidance on the basic catalog mechanism before we start building SQL integration on top. Please see the [Druid catalog issue](https://github.com/apache/druid/issues/12546) for additional details about the goals and design of the catalog.

### Update 10-17

After much debate, we decided to drop the rollup feature from the catalog. It turns out that "rollup" is a feature used for many purposes and more thought is needed to work out how users might want to use the rollup features. Compared to the Catalog proposal, the initial feature won't help with MSQ ingestion of rollup tables: users are empowered to specify on each ingestion how they want those segments to be rolled up.

A new commit pulls out the portions of the code in support of the proposed, but now dropped, `rollup` datasource type.

<hr>

This PR has:
- [X] been self-reviewed.
- [X] has a design document [here](https://github.com/apache/druid/issues/12546).
- [ ] added documentation for new or modified features or behaviors. (Not yet: the functionality is not yet user visible.)
- [X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [X] added integration tests.
- [ ] been tested in a test Druid cluster.

paul-rogers commented:
Build is clean except for one unrelated flaky failure in a Kafka IT. Will rerun later as I'm sure someone will have comments that require code changes and a new build.

rohangarg reviewed (commented):
Thanks a lot for the design issue and changes @paul-rogers! :+1: 
I've currently taken a first pass at the core changes as of now - posting the current review as a checkpoint as I review the remaining part.
Thanks for the in-code explanations and patience as well!

paul-rogers reviewed (commented):
@rohangarg, thanks for your review. Addressed your comments. Also revised the model to remove the "rollup" concept, leaving just the one "datasource" table type.

FrankChen021 commented on core/src/main/java/org/apache/druid/data/input/InputFormat.java:
```diff
@@ -41,15 +41,17 @@
-    @Type(name = "json", value = JsonInputFormat.class),
+    @Type(name = CsvInputFormat.TYPE_KEY, value = CsvInputFormat.class),
+    @Type(name = JsonInputFormat.TYPE_KEY, value = JsonInputFormat.class),
     @Type(name = "regex", value = RegexInputFormat.class),
```
Since other string constants are turned into a static constant in each class, why does this one leave unchanged?

FrankChen021 commented on core/src/main/java/org/apache/druid/data/input/InputSource.java:
```diff
@@ -47,19 +48,22 @@
+    @Type(name = LocalInputSource.TYPE_KEY, value = LocalInputSource.class),
+    @Type(name = HttpInputSource.TYPE_KEY, value = HttpInputSource.class),
+    @Type(name = InlineInputSource.TYPE_KEY, value = InlineInputSource.class),
     @Type(name = "combining", value = CombiningInputSource.class)
```
Same question as above

FrankChen021 commented on core/src/main/java/org/apache/druid/java/util/common/jackson/JacksonUtils.java:
```diff
@@ -89,7 +95,50 @@ public static void writeObjectUsingSerializerProvider(
+   */
+  public static String toString(Object obj)
+  {
+    ObjectMapper jsonMapper = new ObjectMapper();
```
I think we should use `DefaultObjectMapper` instead

FrankChen021 commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/SQLCatalogManager.java:
```diff
@@ -0,0 +1,578 @@
+  }
+
+  @Override
+  public long markDeleting(TableId id)
```
What the lifecycle of a table in catalog? I mean, there's a method to allow us to mark a table as DELETING, but no other methods to turn state from `DELETING` to `ACTIVE`.

FrankChen021 commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/SQLCatalogManager.java:
```diff
@@ -0,0 +1,578 @@
+            int updateCount = handle.createStatement(
+                StringUtils.format(DELETE_TABLE, tableName))
+                .bind("schemaName", id.schema())
+                .bind("name", id.name())
```
So we can delete a table from catalog even its state if `ACTIVE`?

FrankChen021 reviewed (commented):
The change is huge and I just took a short time to look through and left some minor suggestions.

FrankChen021 commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/MetastoreManager.java:
```diff
@@ -0,0 +1,60 @@
+   * If SQL based, return the SQL version of the metastore
+   * connector. Throws an exception if not SQL-based.
+   */
+  SQLMetadataConnector sqlConnector();
```
About this interface and above `isSql`, I think we don't need such a specialized method defined at this interface level. A more appropriate way is to change it as 

```java
MetadataStorageConnector connector();
```
and remove the `isSql` interface. 

The `SQLCatalogManager` checks if the returned connector is instanceof `SQLMetadataConnector` or not.

paul-rogers reviewed (commented):
Thank you @rohangarg and @FrankChen021  for your reviews. I've either updated the code or provided a description for each of our comments.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogListenerResource.java:
```diff
@@ -0,0 +1,126 @@
+    return Response.status(Response.Status.ACCEPTED).build();
+  }
+
+  private Response checkAuth(final HttpServletRequest request)
```
Why not use a ResourceFilter for this as is done for other resource classes?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogListenerResource.java:
```diff
@@ -0,0 +1,126 @@
+    }
+    TableSpec spec = tableSpec.spec();
+    if (CatalogUpdateNotifier.TOMBSTONE_TABLE_TYPE.equals(spec.type())) {
+      listener.deleted(tableSpec.id());
```
Why not have 2 different endpoints; one for update, one for delete?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogListenerResource.java:
```diff
@@ -0,0 +1,126 @@
+      tableSpec = mapper.readValue(inputStream, TableMetadata.class);
+    }
+    catch (IOException e) {
+      return Response.serverError().entity(e.getMessage()).build();
```
Think the error here could be improved. Maybe the error should contain an identifier for the table that was attempted to be synced, and the operation type (delete, update, etc), along with the exception message itself.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+  private enum PostAction
+  {
+    NEW,
+    IFNEW,
```
nit: do we use snake case for enums that contain multiple words?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+    }
+  }
+
+  private Response authorizeTable(TableId tableId, TableSpec spec, final HttpServletRequest req)
```
Why not use resourceFilter?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+      );
+    }
+
+    // Table name can't be blank or have spaces
```
from logic below, it seems that the table can have spaces, just not leading or trailing spaces, is that right? If so we should update this comment.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+    if (actionParam == null) {
+      action = PostAction.NEW;
+    } else {
+      action = PostAction.valueOf(StringUtils.toUpperCase(actionParam));
```
Should be catch IllegalArgumentException  here, in case the actionParam is not a valid action?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+        return Actions.badRequest(
+            Actions.INVALID,
+            StringUtils.format(
+                "Not a valid action: [%s]. Valid actions are new, ifNew, replace, force",
```
nit: can use `PostAction.values()` to list the possible values.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+    // Druid has a fixed set of schemas. Ensure the one provided is valid.
+    Pair<Response, SchemaSpec> result = validateSchema(tableId.schema());
+    if (result.lhs != null) {
+      return result.lhs;
```
Why skip the authorization checks if the response is null?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+   *        {@code ifNew} is the same as the SQL IF NOT EXISTS clause. If {@code new},
+   *        then an error is raised if the table exists. If {@code ifNew}, then
+   *        the action silently does nothing if the table exists. Primarily for
+   *        use in scripts. The other two options are primarily for use in tests.
```
Should replace and force only be available in test mode?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+      case FORCE:
+        return addOrUpdateTableSpec(table);
+      default:
+        throw new ISE("Unknown action.");
```
think the error message could be better here. Maybe same as what you have on line 136.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+    SchemaSpec schema = result.rhs;
+
+    // The schema has to be one that allows table definitions.
+    if (!schema.writable()) {
```
seems that schema can be null here?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+              )
+        );
+      } else {
+        return Actions.okWithVersion(0);
```
Seems weird to return a version of 0 here, why not return the existing version? Maybe the create method called above should have a version that takes `ifNew` and returns the current version if the table already exists?

zachjsh replied:
In response to the potential for information leak, wouldnt the user already be authorized before getting far enough to get a duplicate error? If not they should be, in which case I think we would only leak the version information to an authorized user.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+          .entity(
+              Actions.error(
+                  Actions.DUPLICATE_ERROR,
+                  "The table entry not found or is older than the given version: reload and retry"))
```
Can we not tell between these two cases?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+      long newVersion = catalog.tables().updatePayload(tableId, action);
+      return Actions.okWithVersion(newVersion);
+    }
+    catch (NotFoundException e) {
```
Does it make sense to also throw and catch `OutOfDateException` as is done when force creating a table, so that the user can distinguish between the table not existing and them specifying the incorrect version?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+   * <p>
+   * The operation is done atomically so no optimistic locking is required.
+   *
+   * @param dbSchema
```
would be good to describe the parameters here

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+      @Context final HttpServletRequest req
+  )
+  {
+    if (command == null) {
```
Any reason to not have this validation in `MoveColumn` class itself?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+    if (spec != null) {
+
+      // The given table spec has to be valid for the given schema.
+      if (Strings.isNullOrEmpty(spec.type())) {
```
Should a TableSpec ever not have a type? If not, maybe we should move this validation to the TableSpec class itself?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+          }
+          @SuppressWarnings("unchecked")
+          List<String> hiddenProps = (List<String>) spec.properties().get(AbstractDatasourceDefn.HIDDEN_COLUMNS_PROPERTY);
+          return spec.withProperty(
```
Does it make a difference if the hidden columns are or are not also listed as columns in table column list itself?

zachjsh replied:
makes sense

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+  @Path("/tables/{dbSchema}/{name}/hideColumns")
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response hideColumns(
```
Seems that we are combining both hide and unhide here, should we have separate apis for each?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+    );
+  }
+
+  private static <T extends ColumnSpec> List<T> dropColumns(
```
Should move this into a class as is done with HideColumns and MoveColumns to be consistent.

zachjsh replied:
ah gotcha ok, makes sense.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+   * not impose security on schemas, only tables within schemas.
+   */
+  @GET
+  @Path("/list/schemas/names")
```
do we need to have `list` in the api path? Similar for other apis

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+  @GET
+  @Path("/schemas/{dbSchema}/names")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response listTables(
```
`listTableNames` may be more appropriate name.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+  @GET
+  @Path("/schemas/{dbSchema}/tables")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response listTableDetails(
```
nit: `listTableMetadata` may be more appropriate name, given the data type that is returned.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+  @GET
+  @Path("/schemas/{dbSchema}/tables")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response listTableDetails(
```
Should we have another version of function that allow user to get the table metadata for all tables in any schema they have acces to? Or the schema can be given as an optional query parameter?

zachjsh replied:
makes sense

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+
+  /**
+   * Synchronization request from the Broker for information about a specific table
+   * (datasource). Done on first access to the table by any query. After that, the
```
Why doesnt the coordinator push updates to brokers when table definition is created, why does it need to be triggered by the broker receiving a query for the table? Seems like it could cause slow first queries.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,747 @@
+
+  @POST
+  @Path("/flush")
+  public Response flush(
```
What is the idea for this once implemented? Javadoc would be good, if we know.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/Actions.java:
```diff
@@ -0,0 +1,104 @@
+/**
+ * Helper functions for the catalog REST API actions.
+ */
+public class Actions
```
nice!

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/CatalogAuthorizer.java:
```diff
@@ -0,0 +1,100 @@
+    }
+  }
+
+  public boolean isAuthorized(String resource, String key, Action action, HttpServletRequest request)
```
not used

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/CatalogAuthorizer.java:
```diff
@@ -0,0 +1,100 @@
+    return new ResourceAction(new Resource(name, schema.securityResource()), action);
+  }
+
+  public Action inferAction(HttpServletRequest request)
```
Doesnt seem to be used.

rohangarg commented on server/src/main/java/org/apache/druid/catalog/model/table/ClusterKeySpec.java:
```diff
@@ -0,0 +1,83 @@
+public class ClusterKeySpec
+{
+  private final String expr;
+  private final boolean desc;
```
FWIU, I think clustering is a concept which is currently sorting but might be extended in future to hashing or space filling curves, trees as well. So, I thought that instead of hardcoding an attribute in the spec we could make it extendible. wdyt? 

For column based indexes, I completely agree with you that those can be per-column properties

paul-rogers replied:
I see the point. Do we have a design for the new kinds of clustering at the column level? For example, instead of sorting a key, might we hash it? Or, would we say, "do sort it, but use XYZ sort instead of ABC sort?" How would we represent this in SQL? At present, `CLUSTER BY` looks like a `SORT` spec, I believe.

If we have a design, then I can update this item to match.

If, on the other hand, these other items have to do with _indexing_ of columns (not the segment sort order), then column properties would be the way to go.

zachjsh replied:
Is there any harm is making this extensible even if can be mapped to desc/asc boolean for foreseable future? Would it require a lot of code churn to update this? I agree with @rohangarg that would be good to make this flexible.

paul-rogers replied:
Exactly what would the extension be? Clustering is sorting. Sorting is either ascending or descending. Let's say we had a hash index. Could we have a "hash clustering"? I'm not sure what that means.

If we want to extend, then let's wait for the new innovation. Then, if we, say, provide a hash table as a an alternative to sorting, we provide a `hashing` property that defines the attributes of that alternative, and say that you can provide either clustering or ordering.

If there is something in the works now, then pease point me to the spec and I'll be happy to model it in the catalog.

paul-rogers replied:
One other thought. Note that the asc/desc option is designed to default to the most frequent choice:

```java
  public ClusterKeySpec(
      @JsonProperty("column") String expr,
      @JsonProperty("desc") @Nullable Boolean desc
  )
  {
    this.expr = expr;
    this.desc = desc != null && desc;
  }
```

Let's say we wanted to be SQL compliant and let the user specify `NULLS HIGH` or `NULLS LOW`. We would use the same trick which would be nicely backward compatible:

```java
  public ClusterKeySpec(
      @JsonProperty("column") String expr,
      @JsonProperty("desc") @Nullable Boolean desc,
      @JsonProperty("nullsHigh") @Nullable Boolean nullsHigh
  )
  {
    this.expr = expr;
    this.desc = desc != null && desc;
    this.nullsHigh = nullsHigh != null & nullsHigh;
  }
```

rohangarg reviewed (commented):
Thanks a lot for addressing the comments @paul-rogers! I have responded to the open threads and reviewed the extension's core part from my side. Will finish up the syncing module as well soon.
have also closed some minor comment threads since they seem not important for now.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/CatalogAuthorizer.java:
```diff
@@ -0,0 +1,100 @@
+    authorize(schema.securityResource(), name, action, request);
+  }
+
+  public void authorize(String resource, String key, Action action, HttpServletRequest request)
```
I see that some of the methods in this class are only used within the class, should they be private methods?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/CatalogStorage.java:
```diff
@@ -0,0 +1,148 @@
+import java.util.List;
+
+/**
+ * Facade over the three internal components used to manage the metadata
```
Can you provide pointers to descriptions of what these 3 components are in the javadoc, or describe them explicitly in the javadoc?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/CatalogStorage.java:
```diff
@@ -0,0 +1,148 @@
+    this.authorizer = new CatalogAuthorizer(authorizerMapper);
+  }
+
+  public CatalogAuthorizer authorizer()
```
Dont need this if we make the authorization in Resource class use ResourceFilter.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/CatalogStorage.java:
```diff
@@ -0,0 +1,148 @@
+  }
+
+  @Override
+  public ResolvedTable resolveTable(TableId id)
```
Good to annotate with `@Nullable`

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/HideColumns.java:
```diff
@@ -0,0 +1,102 @@
+import java.util.Objects;
+import java.util.Set;
+
+public class HideColumns
```
javadocs would be good here.

zachjsh replied:
I think you added the javadoc after I reviewed this area, looks good now, thanks

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/MetastoreManager.java:
```diff
@@ -0,0 +1,60 @@
+ */
+public interface MetastoreManager
+{
+  MetadataStorageConnector connector();
```
This is not used

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/MetastoreManager.java:
```diff
@@ -0,0 +1,60 @@
+public interface MetastoreManager
+{
+  MetadataStorageConnector connector();
+  MetadataStorageConnectorConfig config();
```
This is not used.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/MoveColumn.java:
```diff
@@ -0,0 +1,132 @@
+    return revised;
+  }
+
+  private static int findColumn(List<ColumnSpec> columns, String colName)
```
why not use `columns.indexOf(colName)` instead?

zachjsh replied:
ah yeah see that now, thanks for clarifying.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/CatalogManager.java:
```diff
@@ -0,0 +1,167 @@
+
+
+  void register(Listener listener);
+  void createTableDefnTable();
```
Seems a bit weird to have on this interface

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/CatalogManager.java:
```diff
@@ -0,0 +1,167 @@
+ * Performs detailed CRUD operations on the catalog tables table.
+ * Higher-level operations appear elsewhere.
+ */
+public interface CatalogManager
```
Would be good to add any missing javadocs on methods in this interface.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/SQLCatalogManager.java:
```diff
@@ -0,0 +1,578 @@
+                + "  updateTime BIGINT NOT NULL,\n"
+                + "  state CHAR(1) NOT NULL,\n"
+                + "  payload %s,\n"
+                + "  PRIMARY KEY(schemaName, name)\n"
```
Does this need a IF NOT EXIST clause? This seems to be called on lifecycle start, which I believe is called whenever there is a leadership change to this service instance / the cluster comes up. Wont this fail if the table already exists?

zachjsh replied:
ahh. And yes I do see now that we catch the exception at deeper level, though as you said, we dont differentiate between the table existing and not being able to create table / other errors. Thanks for clarifying

paul-rogers reviewed (commented):
Thanks all for the many comments on the REST API. Looks like the REST API should be redesigned. Marking this as a draft for now.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogListenerResource.java:
```diff
@@ -0,0 +1,126 @@
+import java.io.InputStream;
+
+@Path(CatalogListenerResource.BASE_URL)
+public class CatalogListenerResource
```
I think you added the javadoc after I made this comment, from the code excerpt above, seems there was no javadoc before. Thanks for adding!

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/CatalogManager.java:
```diff
@@ -0,0 +1,167 @@
+   * Retryable only if the version is given, and an
+   * {@code OutOfDateException} is thrown.
+   */
+  long update(TableMetadata table, long oldVersion) throws OutOfDateException, NotFoundException;
```
Should we have a separate version that does not take a version, so that this is always updating given a current version, and the other method could always just force update whatever the current version is?

paul-rogers commented:
Here's how we propose to revise the REST API. Comments or suggestions?

Use a form of `{topic}/{path}`, where the topic is the kind of action. This resolves the ambiguity of trying to do it the other way: `{path}/{action}` where there is an ambiguity between paths and actions.

### Configuration-as-Code

CRUD for an entire `TableSpec`, optionally protected by a version.

`POST /resource/tables/{schema}/{table}[?version={n}|overwrite=true|false]`

Create or update a `TableSpec`.

* With no options, the semantics are "create if not exists."
* With a version, the semantics are "update if exists and is at the given version"
* With "overwrite", the semantics are "create or update"

`GET /resource/tables/{schema}/{table}`

`DELETE /resource/tables/{schema}/{table}`

### Editing

Edit operations apply a specific transform to an existing table spec. Because the transform is very specific, there is no need for a version as there is little scope for unintentional overwrites. The edit actions are usable from the Druid console or DDL statements, if/when we add them.

`POST /edit/table/{schema}/{table}`

The payload is a message that says what to change:

* Hide columns (add to the hidden columns list)
* Unhide columns (remove from the hidden columns list)
* Drop columns (remove items from the columns list)
* Move columns
* Update props (merge updated properties with existing)
* Update columns (merge updated columns with existing)

### Retrieval

`GET /names/schemas`

Returns a list of the schema names.

`GET /names/tables/{schema}`

Returns a list of the table names within a schema.

`GET /entry/tables`

Returns a list of the metadata (`TableSpec` plus update dates, etc.) for all tables. 

`GET /entry/tables/{schema}`

Returns a list of the metadata for tables within a schema.

`GET /entry/tables/{schema}/{table}`

Returns the metadata for a single table.

### Sync

`GET /sync/tables`

 Returns a list of `TableSpec` objects for bootstrapping a Broker.

Similar to`GET /entry/tables`, but that message returns metadata.

`POST /sync/delta`

On the broker, receives update (delta) messages from the coordinator. The payload is an object:

* Create/update: table path plus the `TableSpec`
* Delete: table path only

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/SQLCatalogManager.java:
```diff
@@ -0,0 +1,578 @@
+            if (resultIterator.hasNext()) {
+              return resultIterator.next();
+            }
+            return null;
```
nit: should annotate with `@Nullable`

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/SQLCatalogManager.java:
```diff
@@ -0,0 +1,578 @@
+                  .bind("oldVersion", oldVersion)
+                  .execute();
+              if (updateCount == 0) {
+                throw new OutOfDateException(
```
Feels a bit strange to throw this error in both cases; that the table does not exist, or version is wrong. Could do another read to check if it exists?

zachjsh replied:
I think it would be good to distinguish between these 2 error cases. Maybe we can do another DB read to check if the table exists?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/SQLCatalogManager.java:
```diff
@@ -0,0 +1,578 @@
+  public long update(TableMetadata table, long oldVersion) throws OutOfDateException, NotFoundException
+  {
+    if (oldVersion == 0) {
+      return updateUnsafe(table.id(), table.spec());
```
See comment in interface.

paul-rogers commented on server/src/main/java/org/apache/druid/catalog/model/TableMetadata.java:
```diff
@@ -0,0 +1,242 @@
+{
+  public enum TableState
+  {
+    ACTIVE("A"),
```
At present there are two states. I can't imagine adding many more, most likely zero. But, since there is some chance we will want another state, and since this is a DB field, we use the old "encode an enum as a single character" trick. The column itself is (or should be) a `CHAR(1)`. As a result, there an be no states with more than one letter.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/SQLCatalogManager.java:
```diff
@@ -0,0 +1,578 @@
+            public Long withHandle(Handle handle) throws OutOfDateException
+            {
+              long updateTime = System.currentTimeMillis();
+              int updateCount = handle.createStatement(
```
Should we allow update for a table that is in non-active state? Same for other update methods.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/SQLCatalogManager.java:
```diff
@@ -0,0 +1,578 @@
+                sendUpdate(id);
+                return updateTime;
+              }
+              catch (RuntimeException e) {
```
Should we catch any exception / throwable here and do rollback?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/SQLCatalogManager.java:
```diff
@@ -0,0 +1,578 @@
+              long updateTime = System.currentTimeMillis();
+              int updateCount = handle.createStatement(
+                  StringUtils.format(UPDATE_DEFN_SAFE, tableName))
+                  .bind("schemaName", id.schema())
```
I dont believe that the schema or name of table can change can it? Since doing so would create a new entry in table. Given this maybe we shouldnt bind the schemaName nor name in the  update clause here. Same for other update methods.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/CatalogManager.java:
```diff
@@ -0,0 +1,202 @@
+   * Delete the table record for the given ID. Essentially does a
+   * "DELETE IF EXISTS". There is no version check. Delete should be
+   * called only when there are no segments left for the table: use
+   * {@link #markDeleting(TableId)} to indicates that the segments are
```
To clarify, is the plan for calling the markDeleting method to trigger segment deletion for the underlying datasource?

zachjsh replied:
Makes sense. Thanks for clarifying

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/CatalogManager.java:
```diff
@@ -0,0 +1,202 @@
+  /**
+   * Returns a list of the ids (schema, name) of all tables in the catalog.
+   */
+  List<TableId> list();
```
nit: think `listIds` would be more clear.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/CatalogManager.java:
```diff
@@ -0,0 +1,202 @@
+  /**
+   * Returns a list of the names of all tables within the given schema.
+   */
+  List<String> list(String dbSchema);
```
nit: think `listNames` would be more clear.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/CatalogManager.java:
```diff
@@ -0,0 +1,202 @@
+   * Returns a list of the table metadata for all tables within the given
+   * schema.
+   */
+  List<TableMetadata> listDetails(String dbSchema);
```
nit: think `listMetadata` would be more clear.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/sync/CatalogUpdateNotifier.java:
```diff
@@ -0,0 +1,121 @@
+  @Override
+  public void deleted(TableId tableId)
+  {
+    TableMetadata spec = TableMetadata.newTable(tableId, TABLE_TOMBSTONE);
```
If the listener wanted to delegate proper deletion based on the type of the table, how would they do this with this special tombstone type? For example say some type of table's deletion process is different than another.

paul-rogers replied:
The notification mechanism has changed to use events, so no more tombstones. The deletion event here is that the table was deleted from the catalog DB. This can happen because the user made a mistake, and just deletes an unwanted entry. Or, more formally, it would be the last step in a deletion process outlined in an earlier comment. This event should trigger nothing other than removal from a cache on the broker.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/sync/CacheNotifier.java:
```diff
@@ -0,0 +1,118 @@
+
+  public void stop()
+  {
+    exec.shutdownNow();
```
might be good to log if we have any updates left before shutting down, if this is a non-graceful shutdown

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/sync/CatalogUpdateNotifier.java:
```diff
@@ -0,0 +1,121 @@
+    this.notifier = new CacheNotifier(
+        CALLER_NAME,
+        sender);
+    catalog.register(this);
```
Should only the leader be able to register itself and push updates to listeners? If so, what is controlling that?

paul-rogers replied:
How do we handle this issue generically? I'd kind of hoped that the Coordinator itself will reject messages if it is the follower.

zachjsh replied:
Take a look at the `TaskMaster` class, you'll see that it does things like startup the taskRunner on on the leader when it comes up, and shutdown on follower when is loses leadership. Maybe can follow that sort of pattern? Not sure if needed. There is logic that routes the coordinator request to leader, but not sure if there is anyway for that to get out of sync with the leadership election, and route request to wrong host.

zachjsh replied:
Yeah maybe this isnt needed. Will just leave up here, but feel free to resolve, if you feel we dont need too.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/sync/CachedMetadataCatalog.java:
```diff
@@ -0,0 +1,229 @@
+  {
+    private final TableMetadata table;
+
+    protected TableEntry(SchemaSpec schema, TableMetadata table)
```
schema parameter doesn't seem to be used

zachjsh commented:
> Here's how we propose to revise the REST API. Comments or suggestions?
> 
> Use a form of `{topic}/{path}`, where the topic is the kind of action. This resolves the ambiguity of trying to do it the other way: `{path}/{action}` where there is an ambiguity between paths and actions.
> ### Configuration-as-Code
> 
> CRUD for an entire `TableSpec`, optionally protected by a version.
> 
> `POST /resource/tables/{schema}/{table}[?version={n}|overwrite=true|false]`
> 
> Create or update a `TableSpec`.
> 
>     * With no options, the semantics are "create if not exists."
> 
>     * With a version, the semantics are "update if exists and is at the given version"
> 
>     * With "overwrite", the semantics are "create or update"
> 
> 
> `GET /resource/tables/{schema}/{table}`
> 
> `DELETE /resource/tables/{schema}/{table}`
> ### Editing
> 
> Edit operations apply a specific transform to an existing table spec. Because the transform is very specific, there is no need for a version as there is little scope for unintentional overwrites. The edit actions are usable from the Druid console or DDL statements, if/when we add them.
> 
> `POST /edit/table/{schema}/{table}`
> 
> The payload is a message that says what to change:
> 
>     * Hide columns (add to the hidden columns list)
> 
>     * Unhide columns (remove from the hidden columns list)
> 
>     * Drop columns (remove items from the columns list)
> 
>     * Move columns
> 
>     * Update props (merge updated properties with existing)
> 
>     * Update columns (merge updated columns with existing)
> 
> 
> ### Retrieval
> 
> `GET /names/schemas`
> 
> Returns a list of the schema names.
> 
> `GET /names/tables/{schema}`
> 
> Returns a list of the table names within a schema.
> 
> `GET /entry/tables`
> 
> Returns a list of the metadata (`TableSpec` plus update dates, etc.) for all tables.
> 
> `GET /entry/tables/{schema}`
> 
> Returns a list of the metadata for tables within a schema.
> 
> `GET /entry/tables/{schema}/{table}`
> 
> Returns the metadata for a single table.
> ### Sync
> 
> `GET /sync/tables`
> 
> Returns a list of `TableSpec` objects for bootstrapping a Broker.
> 
> Similar to`GET /entry/tables`, but that message returns metadata.
> 
> `POST /sync/delta`
> 
> On the broker, receives update (delta) messages from the coordinator. The payload is an object:
> 
>     * Create/update: table path plus the `TableSpec`
> 
>     * Delete: table path only

Thanks for updated proposal @paul-rogers . One thing I think would be good is if in the URI path, when specifying a table within a specific schema, the `schema` should come before `table` in the path, as is denotes a hierarchy, tables within a schema, not the other way around. Anyway that we can do this?

paul-rogers commented:
> One thing I think would be good is if in the URI path, when specifying a table within a specific schema, the schema should come before table in the path

As it turns out, the schema does come before the table:

`POST /resource/tables/{schema}/{table}[?version={n}|overwrite=true|false]`

I suspect the confusing bit is the "/resource/tables". The thought here was that, at present, the catalog has only tables. Most DB allow user-defined schemas. We may want to add connections for things like S3, Kafka, etc. So, the thought was we'd have a variety of "resources". Each with some naming convention. So, in the future (not now):

`POST /resource/schemas/{schema}`
`POST /resource/connections/{conn}`

Etc.

Here, we could simplify: the `resource` part could be removed:

`POST /tables/{schema}/{table}`
`POST /schemas/{schema}`
`POST /connections/{conn}`

(Everything here is simplified, BTW, there is a common prefix which I'm omitting.)

> Maybe the schema can be included in the request payload

That is one solution. To add, `POST /tables`. The problem is, the result is asymmetric on get: `GET /tables` might return everything, so one would do `GET /tables/{schema}/{table}`. Plus, the content would differ: to create we provide a name, but to get, we don't need a response with the name because we already have the name.

Then, there is the update ambiguity: `POST /tables/{schema}/{name}` says which table we want to update. If the name also appears in the request, then we (and the user) would have to ensure that they match. I suppose we could do update as `POST /tables` with the name in the body...

A final comment is that the present design says that the name is the place you store your table spec: it isn't an attribute of the spec. This means I can post the same spec under multiple names: one for dev, another for test, and another for prod. (Since Druid doesn't allow user-defined namespaces, the best thing is "dev_events", "test_events" and "event" for dev, test and prod.) If the name were in the spec, then the spec would have to be modified for each use. (And, the DB record would store the name twice: once in the key field, another in the spec, resulting in redundancy and another thing to verify on every update.)

The existing, and proposed, designs allow the same spec format for create, update and read. It allows the same spec to be posted to dev, test and prod tables. I think we want to keep each of these features. That said, I'm open to revisions about _how_ we provide those features.

> its a bit awkward that the table is specified as a resource in some apis, and as a entry in others

The difference in "themes" was due to the `/schemas/{schema}/{table}`/ `schemas/{schema}/{operation}` ambiguity if we do the obvious solution and try to use a common base for both. But, if we're OK with `/schemas/{schema}/{operation}` and `/tables/{schema}/{table}/{operation}`, then we can almost, but not quite, combine resources and entries. 

For tables:

* `POST /tables/{schema}/{table}[?version={n}|overwrite=true|false]` add/update a table
* `POST /tables/{schema}/{table}/edit` "edit" (incremental update) a table
* `GET /tables/{schema}/{table}` Get the table spec for a table (same object a for create/update)
* `GET /tables/{schema}/{table}/metadata` Get the table metadata (name, update date, state, spec, etc.)
* `GET /tables` Get metadata for all tables in all schemas

For schemas:

* `POST /schemas/{schema}` Create a schema (not yet supported!)
* `GET /schemas/{schema}` Get metadata for a schema (not yet supported!)
* `GET /schemas/{schema}/names` Get the names of tables within the schema
* `GET /schemas/{schema}/tables` Get the metadata for each table in the schema

In the above, however, there is no good way to get the names of all tables in all schemas: `GET /schemas/names` won't work (ambiguous). `GET /schemas` won't work (would imply getting the metadata (contents) for all schemas.

This is the "trying to be too clever" issue that made the original API a bit awkward: had to do some song and dance to work around ambiguities.

The proposal in the earlier message resolves these issues by saying _what you want to do_, then saying, _what you want to do it on_. That way, to get names:

* `GET /names/schemas` says to get all schema names
* `GET /names/tables` says to get all table names in all schemas
* `GET /names/schema/{schema}` says to get all table names in the given schema

(The above is a refinement of the earlier proposal.)

The same pattern is then repeated for metadata with `entries`.

Again, I think we need the lists of names, and the lists of contents. We need it for the whole system, for everything in a schema, and for a single table. Again, I'm open to other ways of accomplishing the goals.

Ideas?

paul-rogers commented:
PR has been updated with a large set of changes based on review comments:

* Revised the REST API per the design outlined above.
* Replaced the ad-hoc ways to "edit" a table with a new API based on "edit commands."
* Replaced sync protocol with one based on "sync events"
* Revised the UTs and ITs to match

paul-rogers commented on server/src/main/java/org/apache/druid/catalog/model/ColumnSpec.java:
```diff
@@ -0,0 +1,141 @@
+{
+  private final String type;
+  private final String name;
+  private final String sqlType;
```
I agree that `ColumnType` is the actual Druid type. The `ColumnType` is the storage type, which is limited. For example, all complex types are represented as `UKNOWN_COMPLEX`, which is the physical, not user view. Over time, we'll want an actual type name for a complex type: something like `IP_ADDRESS` rather than just "complex". The specific names are an active point of discussion. Also, a `__time` is a `LONG` internally, but it is a `TIMESTAMP` semantically.

We do limit the set of SQL types: only those that directly map to Druid types.

zachjsh replied:
I think the descriptions of the fields given here should go directly into javadoc so that the documentation lives forever in the code, and not only in a comment thread.

paul-rogers commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,774 @@
+  {
+
+    TableDefnRegistry tableRegistry = catalog.tableRegistry();
+    return incrementalUpdate(
```
The REST API has been completely revised. Please take a look.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,586 @@
+   * Create or update a new table containing the given table specification.
+   * Supports three use cases:
+   * <ul>
+   * <li>"create if not exists": default use case with no options.</li>
```
Looking at code below, it seems that if no options are given, that the default would be that if a table with same name already exists, we will throw the duplicatKeyException on line 161. Isnt this not `create if not exists`, but rather just `create`?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,586 @@
+  @Path("/schemas/{schema}/tables/{name}/edit")
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response editTable(
```
would be good to add more details about the different edit operations here, and their semantics, or maybe just a pointer the the different edit classes, which themselves describe semantics.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,586 @@
+          return Response.ok().entity(catalog.schemaRegistry().names()).build();
+        case PATH_FORMAT:
+          return listTablePaths(req);
+        case METADATA_FORMAT:
```
Should this also have  a status format as the api below does?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,586 @@
+   * which will probably differ from the list of actual tables. For example, for
+   * the read-only schemas, there will be no table definitions.
+   *
+   * @param dbSchema The Druid schema to query. The user must have read access.
```
Good to describe the format param here too.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,586 @@
+    }
+  }
+
+  private void validateTableSpec(SchemaSpec schema, String name, TableSpec spec) throws CatalogException
```
name parameter not used.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/MoveColumn.java:
```diff
@@ -0,0 +1,79 @@
+  @JsonCreator
+  public MoveColumn(
+      @JsonProperty("column") final String column,
+      @JsonProperty("where") final Position where,
```
Should we fail construction of this class if anchor is null, and position is one of BEFORE or AFTER?

zachjsh replied:
I see we are doing the validation in the TableEditor class. I think its best to move the validation as high up in the stack as possible, and then users can assume that the object has already been validated. What do you think?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/TableEditor.java:
```diff
@@ -0,0 +1,369 @@
+    this.editRequest = editRequest;
+  }
+
+  public long go() throws CatalogException
```
what does return value indicate? Good to add javadoc

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/TableEditor.java:
```diff
@@ -0,0 +1,369 @@
+import java.util.Map;
+import java.util.Set;
+
+public class TableEditor
```
Good to add javadoc.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/TableEditRequest.java:
```diff
@@ -0,0 +1,103 @@
+    @Type(name = "updateColumns", value = TableEditRequest.UpdateColumns.class),
+    @Type(name = "moveColumn", value = MoveColumn.class),
+})
+public class TableEditRequest
```
Good to include javadoc, with description of each operation type.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/TableEditor.java:
```diff
@@ -0,0 +1,369 @@
+  {
+    final TableSpec existingSpec = table.spec();
+    final TableDefn defn = resolveDefn(existingSpec.type());
+    final List<ColumnSpec> revised = defn.mergeColumns(existingSpec.columns(), updates);
```
How do we protect a user from dropping a column, and a column with the same name back, after. In this case, if the table is a Druid datasource,  the data for the previous column may not have been deleted, in which cause querying the column will return old data, whereas the user would likely assume the data should be null. Should we document this, or store a set of columns that have been previously deleted?

zachjsh replied:
makes sense.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/TableEditor.java:
```diff
@@ -0,0 +1,369 @@
+    if (CollectionUtils.isNullOrEmpty(existingColumns)) {
+      return null;
+    }
+    Set<String> drop = new HashSet<String>(toDrop);
```
nit: Explicit type argument String can be replaced with <>

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/CatalogManager.java:
```diff
@@ -0,0 +1,153 @@
+   * Does not cascade deletes yet. Eventually, should delete all entries
+   * for the table.
+   *
+   * @return {@code true} if the table exists and was deleted,
```
delete has not longer returns boolean

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java:
```diff
@@ -0,0 +1,586 @@
+   * @param spec The new table definition.
+   * @param version the expected version of an existing table. The version must
+   *        match. If not (or if the table does not exist), returns an error.
+   * @param overwrite if {@code true}, then overwrites any existing table.
```
Would be good to indicate that this property is ignored, if version is specified.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/CatalogManager.java:
```diff
@@ -0,0 +1,153 @@
+   * @return          the update timestamp (version) of the updated record
+   */
+  long updateProperties(TableId id, TableTransform transform) throws CatalogException;
+  long updateColumns(TableId id, TableTransform transform) throws CatalogException;
```
nit: similar javadoc for updateColumns would be good.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/SQLCatalogManager.java:
```diff
@@ -0,0 +1,777 @@
+            StringUtils.format(
+                CREATE_TABLE,
+                tableName,
+                connector.getPayloadType(),
```
BLOB type has max size of 64KB, I feel that we can easily overrun this for a table with thousands of columns. `MEDIUMBLOB` may be better

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/ColumnSpec.java:
```diff
@@ -0,0 +1,130 @@
+
+/**
+ * Specification of table columns. Columns have multiple types
+ * represented via the type field.
```
Would be good to describe the fields of this class in the javadoc. For example `type` refers to `dimension`, or `measure`, not data type.

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/SQLCatalogManager.java:
```diff
@@ -0,0 +1,777 @@
+
+  private static final String REPLACE_SPEC_STMT =
+      "UPDATE %s\n SET\n" +
+      "  tableType = :tableType,\n" +
```
should  a table be able to change types?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/sync/ListeningNodeSupplier.java:
```diff
@@ -0,0 +1,67 @@
+ * likely to be low, so creating the list each time is fine. If this code
+ * is used for high-speed updates, then caching would be desirable.
+ */
+public class ListeningNodeSupplier implements Supplier<Iterable<DruidNode>>
```
Nice! This could be useful elsewhere in Druid codebase. But fine to leave here now, not sure of a good place for it.

zachjsh commented on extensions-core/druid-catalog/src/test/java/org/apache/druid/server/http/catalog/CatalogResourceTest.java:
```diff
@@ -0,0 +1,697 @@
+  }
+
+  @Test
+  public void testCreate()
```
would be nice to split, so that each test is one function, easier to follow the tests, and what is tested. Also with this, the names of the test functions can clearly communicate what is being tested: for example:


```
@Test
  public void given_emptySchemaName_when_postTable_then_throwBadRequest()
  {
    final String tableName = "create";
    TableSpec dsSpec = TableBuilder.datasource(tableName, "P1D").buildSpec();

    // Blank schema name: infer the schema.
    Response resp = resource.postTable("", tableName, dsSpec, 0, false, postBy(CatalogTests.SUPER_USER));
    assertEquals(Response.Status.BAD_REQUEST.getStatusCode(), resp.getStatus());
}
```

If you agree, can do in follow up.

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/CatalogUtils.java:
```diff
@@ -0,0 +1,147 @@
+
+public class CatalogUtils
+{
+  public static List<String> columnNames(List<ColumnSpec> columns)
```
This seems to only be used in tests, should it be used in main code too?

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/ColumnDefn.java:
```diff
@@ -0,0 +1,110 @@
+      this.spec = spec;
+    }
+
+    public ColumnDefn defn()
```
this is never used

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/ColumnDefn.java:
```diff
@@ -0,0 +1,110 @@
+
+import java.util.List;
+import java.util.Map;
+
```
Good to add javadoc on this main class, as well as on the `merge` method.

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/ColumnDefn.java:
```diff
@@ -0,0 +1,110 @@
+  public void validateScalarColumn(ColumnSpec spec)
+  {
+    Columns.validateScalarColumn(spec.name(), spec.sqlType());
+    if (Columns.isTimeColumn(spec.name())) {
```
it seems that the `validateScalarColumn` call on line 99 above is already performing this check, and failing, is the code below reachable?

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/Columns.java:
```diff
@@ -0,0 +1,105 @@
+  public static final String DOUBLE = "DOUBLE";
+  public static final String TIMESTAMP = "TIMESTAMP";
+
+  public static final Set<String> NUMERIC_TYPES = ImmutableSet.of(BIGINT, FLOAT, DOUBLE);
```
not used.

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/Columns.java:
```diff
@@ -0,0 +1,105 @@
+import java.util.Map;
+import java.util.Set;
+
+public class Columns
```
Should we add a private constructor to this class, since it seems it should just be used as a static utility class?

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/ModelProperties.java:
```diff
@@ -0,0 +1,305 @@
+    String name();
+    String typeName();
+    void validate(Object value, ObjectMapper jsonMapper);
+    Object merge(Object existing, Object update);
```
Should merge also return type `T`?

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/ObjectDefn.java:
```diff
@@ -0,0 +1,138 @@
+   * property is known, use the column definition to merge the values. Else, the
+   * update replaces any existing value.
+   * <p>
+   * This method does not validate the properties, except as needed to do a
```
Is there ever a case where the caller would not want to ultimately validate the merged properties? Any reason not to validate here?

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/ParameterizedDefn.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.List;
+import java.util.Map;
+
+public interface ParameterizedDefn
```
Good to add javadocs

zachjsh replied:
Javadoc on this class + on public methods would be helpful.

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/TableDefn.java:
```diff
@@ -0,0 +1,176 @@
+   * Merge the set of columns from an existing spec and an update.
+   * Columns are matched by name. If the column exists, then it is updated. If
+   * the column does not exist, then the new column is appended to the existing
+   * list. This merge operation cannot remove columns or change order.
```
What is the reason not to allow removal or changing order?

zachjsh replied:
Didn't realize that update here could be just a subset of the columns, maybe I missed that in the java doc.

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/TableDefnRegistry.java:
```diff
@@ -0,0 +1,87 @@
+
+import java.util.Map;
+
+public class TableDefnRegistry
```
Good to add javadoc.

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/TableDefnRegistry.java:
```diff
@@ -0,0 +1,87 @@
+  private final Map<String, TableDefn> defns;
+  private final ObjectMapper jsonMapper;
+
+  public TableDefnRegistry(
```
Should this constructor be private?

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/TableId.java:
```diff
@@ -0,0 +1,113 @@
+
+  @JsonCreator
+  public TableId(
+      @JsonProperty("schema") String schema,
```
Should we add validation that checks that both of these parameters are nonnull?

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/TableMetadata.java:
```diff
@@ -0,0 +1,300 @@
+    );
+  }
+
+  public static TableMetadata newSegmentTable(
```
not used

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/TableMetadata.java:
```diff
@@ -0,0 +1,300 @@
+      @JsonProperty("creationTime") long creationTime,
+      @JsonProperty("updateTime") long updateTime,
+      @JsonProperty("state") TableState state,
+      @JsonProperty("spec") TableSpec spec)
```
Should any of these parameters be validated in anyway, notnull, etc?

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/TableSpec.java:
```diff
@@ -0,0 +1,143 @@
+    return new TableSpec(type, properties, columns);
+  }
+
+  public TableSpec withProperty(String key, Object value)
```
not used

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/table/AbstractDatasourceDefn.java:
```diff
@@ -0,0 +1,144 @@
+   * If unset, then the system setting is used.
+   */
+  public static final String TARGET_SEGMENT_ROWS_PROPERTY = "targetSegmentRows";
+  public static final String CLUSTER_KEYS_PROPERTY = "clusterKeys";
```
Good to describe what this is

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/table/AbstractDatasourceDefn.java:
```diff
@@ -0,0 +1,144 @@
+    return DatasourceDefn.TABLE_TYPE.equals(tableType);
+  }
+
+  public static boolean isDatasource(ResolvedTable table)
```
not used

paul-rogers replied:
Actually, is used in `TableEditor`.

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/table/ClusterKeySpec.java:
```diff
@@ -0,0 +1,78 @@
+
+import java.util.Objects;
+
+public class ClusterKeySpec
```
javadoc would be good

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/table/ExternalTableDefn.java:
```diff
@@ -0,0 +1,273 @@
+  {
+    public static final String FORMAT_PROPERTY = "format";
+
+    private Map<String, InputFormatDefn> formats;
```
could be made final

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/table/ExternalTableDefn.java:
```diff
@@ -0,0 +1,273 @@
+ * Tables which are parameterized implement that interface to expose
+ * methods defined here.
+ */
+public abstract class ExternalTableDefn extends TableDefn
```
Any abstract functions that extending classes need to implement would be good to javadoc, so that users know what the contract is.

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/table/ExternalTableDefn.java:
```diff
@@ -0,0 +1,273 @@
+    return convertToExtern(revised);
+  }
+
+  public static boolean isInputTable(ResolvedTable table)
```
not used

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/table/HttpTableDefn.java:
```diff
@@ -0,0 +1,190 @@
+  public static final String PASSWORD_PROPERTY = "password";
+  public static final String PASSWORD_ENV_VAR_PROPERTY = "passwordEnvVar";
+  public static final String URIS_PROPERTY = "uris";
+  public static final String URIS_PARAMETER = "uris";
```
why are URIS both a property and parameter?

paul-rogers replied:
Because it is? I can specify it _either_ in the catalog (property) or in a SQL table statement (parameter). Not all properties are parameters and visa-versa. In this case, `uris` is both.

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/table/HttpTableDefn.java:
```diff
@@ -0,0 +1,190 @@
+public class HttpTableDefn extends FormattedExternalTableDefn implements ParameterizedDefn
+{
+  public static final String TABLE_TYPE = HttpInputSource.TYPE_KEY;
+  public static final String URI_TEMPLATE_PROPERTY = "template";
```
Can you describe these properties. and how the user should go about setting them?

paul-rogers replied:
I rather expected to provide that info in the docs. It already exists in the issue describing the feature. Us developer-types get the correct, up-to-date info by looking at the code.

zachjsh replied:
Sure, describing in docs fine too.

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/table/InputFormats.java:
```diff
@@ -0,0 +1,301 @@
+       // hasHeaderRow is required, even though we don't infer headers.
+      jsonMap.put("hasHeaderRow", false);
+      jsonMap.put("findColumnsFromHeader", false);
+            // Column list is required. Infer from schema.
```
Does this mean that the order of table columns must match the order of columns in the input? This seems like pretty big limitation.

zachjsh replied:
Didnt realize that MSQ does not support inferring columns from headers. When it does, maybe should revisit this.

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/table/LocalTableDefn.java:
```diff
@@ -0,0 +1,123 @@
+import java.util.List;
+import java.util.Map;
+
+public class LocalTableDefn extends FormattedExternalTableDefn implements ParameterizedDefn
```
javadoc

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/table/HttpTableDefn.java:
```diff
@@ -0,0 +1,190 @@
+public class HttpTableDefn extends FormattedExternalTableDefn implements ParameterizedDefn
+{
+  public static final String TABLE_TYPE = HttpInputSource.TYPE_KEY;
+  public static final String URI_TEMPLATE_PROPERTY = "template";
```
is `template` too generic? Maybe `urisTemplate`?

paul-rogers replied:
This is one of those context questions. To you, reading this file amid zillions of others, the term "template" is too generic. To a user, focused on the HTTP input source, there is only one thing that could be templated: the URL. Since these values also appear in SQL, shorter names is a convenience for the user:

```SQL
... FROM (TABLE(http(template => 'http://foo.bar/{}.html')))
```
vs.
```SQL
... FROM (TABLE(http(httpInputSourceURLTemplate => 'http://foo.bar/{}.html')))
```

That said, changed it to `uriTemplate`. Three more letters to type and a shift key to hit...

zachjsh commented on server/src/main/java/org/apache/druid/catalog/model/table/TableBuilder.java:
```diff
@@ -0,0 +1,269 @@
+ * table. Not useful outside of tests since Druid code should not create
+ * table objects except generically from specs provided by the user.
+ */
+public class TableBuilder
```
Why not move to test folder?

paul-rogers replied:
Turns out this is needed in the ITs. Would be a hassle to create a test jar. Easiest to just leave where it is.

zachjsh commented on server/src/main/java/org/apache/druid/metadata/storage/derby/DerbyMetadataStorage.java:
```diff
@@ -55,6 +54,26 @@ public void start()
       throw new RuntimeException(e);
     }
+
+    // It takes a while for the Derby server to start in another
```
nice!

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/sync/CachedMetadataCatalog.java:
```diff
@@ -84,14 +97,14 @@ protected TableEntry(TableMetadata table)
     private final SchemaSpec schema;
     private long version = NOT_FETCHED;
-    private final ConcurrentHashMap<String, TableEntry> cache = new ConcurrentHashMap<>();
+    private final Map<String, TableEntry> cache = new TreeMap<>();
```
Seems like using treeMap here, and adding sychronization at method level really lowers the granularity of concurrent access; now we have entire functions being synchronized, thought better to use concurrentMap, so that updates to each table are synchronized, but not each table can be updated independently and concurrently with another. I notice that you had the `update`  method in this class as synchronized before even using treeMap, why is this?

zachjsh commented on extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/sync/CachedMetadataCatalog.java:
```diff
@@ -0,0 +1,331 @@
+ * that schema is marked as not existing and won't be fetched again.
+ *
+ * The cache is updated via an update facility which either flushes
+ * the cache (crude) or listens to the base catalog for updates and
```
Sounds good, will just leave this here as a reminder for next pr

zachjsh reviewed (approved):
Thanks for making all the changes!! Once we get the tests fix, think this is good to go.
